spark/src/main/scala/org/apache/spark/sql/delta/actions/TableFeatureSupport.scala

```diff
@@ -254,6 +317,32 @@ trait TableFeatureSupport { this: Protocol =>
     }
   }
 
+  /**
+   * Remove writer feature from protocol. To remove a writer feature we only need to
+   * remove it from the writerFeatures set.
+   */
+  private[delta] def removeWriterFeature(targetWriterFeature: WriterFeature): Protocol = {
+    require(targetWriterFeature.isRemovable)
+    copy(writerFeatures = writerFeatures.map(_ - targetWriterFeature.name))
+  }
+
+  /**
+   * Remove feature wrapper for removing either Reader/Writer or Writer features. We assume
+   * the feature exists in the protocol. There is a relevant validation at
+   * [[AlterTableRemoveFeatureDeltaCommand]].
+   */
+  def removeFeature(targetFeature: TableFeature): Protocol = targetFeature match {
+    case f: ReaderWriterFeature =>
+      throw new UnsupportedOperationException(
+        "ReaderWriter feature removal is currently not supported")
+    case f: WriterFeature =>
+      removeWriterFeature(f)
+    // Legacy features not currently supported.
+    case _ =>
+      throw new UnsupportedOperationException(
+        "Feature removal is not supported for this feature type")
```

Comment:
Don't these need error classes since they're user facing?

Also, shouldn't the error message actually include the offending feature name?

Reply:
Added an error class for that message.